### DISCUSSION TITLE
Any tips to increase performance?

### DISCUSSION THREAD

**Discussion by matthewjosephtaylor** [timocov/dts-bundle-generator#254] [General]
I notice this tool (which is awesome!) runs a bit slower than I would have anticipated.

I'm running it on a large monorepo with many workspace sub-projects, so I'm not to astonished it takes a while...but still. Like a minute or two.

If there is a known 'hotspot' in the code I'd love to take a crack at cooling it down. 

Otherwise if there are any command line tweaks I'd be interested in those as well.

Thanks for such a useful tool (just what I needed, and I can see by the codebase that it must have taken quite some effort to code up)!

**Comment by timocov** [comment]
Hi there,

the tool relies on the typescript compiler a lot, and from my experience the most of the execution time is taken by compiling the code rather than the tool's internals.

What time it takes to compile your project? What do you use as an input for the tool - dts file(s) or ts? Do you compile your code in the pipeline/build steps somewhere before running the tool?

**Reply by matthewjosephtaylor** [reply]
Ah, that explains it!

I'm currently building each workspace 'sub-project' independently (ts->js via tsc -b) as needed.

So each sub-project has a /dist folder with 'compiled' js and .d.ts files inside of it.

Then rolling them all up via esbuild (sub-project dist/js -> singlle-file-js).

So it makes sense. Each sub-project isn't huge but there are like a dozen or so of them and so are adding up when done all at once.

So perhaps if I broke out the .d.ts files independent of the code and ran dts-bundler-generator on those directly (instead of the .d.ts + js) that might be faster.  

I'll give that a shot and report back on findings....

**Reply by timocov** [reply]
Not sure if it would help, if you share the code I'd probably look closer and maybe can provide some suggestions. The reason why I asked about dts files is that the tool compiles ts files into dts first and then compiles dts files once again, but if at the first step the program contains only dts files (i.e. your input is dts file and any file that is used is dts) then the tool will skip that second compilation which might improve the performance a bit (you can try to play with `--verbose` flag to see more messages, but frankly I haven't used them for a while so not sure if they can help).

**Comment by rechl** [comment]
Hi @timocov 

Some data for this thread, in case it's useful — the bottleneck in our project turned out
to be the tool's internals rather than the TypeScript compiler.

Setup: dts-bundle-generator 9.5.1, TypeScript 5.9.3, Node 24.14. The entry file has ~55
export statements pulling types out of a `src/` tree of ~486 `.ts` files. Run with
`--no-check --project tsconfig.json`. That single entry took **78 seconds**.

A `--cpu-prof` run attributes 74.7s of 79.3s total (94%) to a single function:

```
  74.73s  isSymbolUsedBySymbolImpl  [types-usage-evaluator.js:18]
   3.45s  (program)
   0.23s  (garbage collector)
   0.18s  (anon)                    [bundle-generator.js:585]
   ...everything else < 0.1s
```

TypeScript compilation is essentially free here by comparison, which surprised me given the
earlier note in this thread that most of the time usually goes into compilation. I suspect
it depends on the shape of the entry point: the cost seems to scale with (number of exported
symbols x size of the symbol graph), so it stays invisible until the export surface grows.

Bisecting the entry file matches that. Each half is cheap on its own, together they cost more
than the sum:

| entry                                | time |
| ------------------------------------ | ---- |
| trivial `export type X = { a: string }` | 0s  |
| first half of the exports            | 0s   |
| second half of the exports           | 26s  |
| both halves (the real entry file)    | 78s  |

The reason is documented in the code itself — `isSymbolUsedBySymbolImpl` deliberately never
caches a negative result, because with a shared `visitedSymbols` set a `false` can be an
artifact of traversal order rather than a fact about the graph:

```js
// note that we can't save negative result here because it might be not a final one
// because we might ended up here because of `visitedSymbols.has(symbol)` check above
```

That reasoning is correct for a per-query search. But the ambiguity disappears if the question
is turned around: instead of asking "is `to` reachable from `from`?" per query, compute the
full set of symbols reachable from `from` once and answer from it. The visited set then *is*
the result — complete, and independent of query order — so negatives become as trustworthy as
positives.

I tried that locally (iterative traversal over the same `nodesParentsMap` edges, one cached
`Set` per source symbol):

- the entry above: **78s -> 0.8s**
- our full build of 8 entry points: **~10 minutes -> 3.9s**
- output is **byte-identical** in all 8 cases

The tradeoff is memory: one reachable set per symbol instead of individual pairs. For a
general-purpose library that may well be the wrong default, and a variant that walks backwards
once per exported symbol would get the same complexity with only one set alive at a time.

I'm running this as a local patch and I'm not blocked, so this is purely FYI — no need to act
on it. If a fix along these lines would be welcome I'm happy to open a PR against
`src/types-usage-evaluator.ts` with your test suite green and a case covering cyclic types
(the shape from #314, where the old visited-set logic was sensitive). Just say the word, and
otherwise please ignore this. Thanks for the tool.
